langchain/src/chat_models/googlepalm.ts

```diff
@@ -0,0 +1,242 @@
+
+  topK?: number = undefined; // default value chosen based on model
+
+  examples: protos.google.ai.generativelanguage.v1beta2.IExample[] = [];
+
+  apiKey?: string;
+
+  private client: DiscussServiceClient;
+
+  constructor(fields?: GooglePalmChatInput) {
+    super(fields ?? {});
+
+    this.model = fields?.model ?? this.model;
+    if (this.model && !this.model.startsWith("models/")) {
+      throw new Error(
+        "`model` value must follow the pattern - `models/{model}`"
+      );
+    }
+
+    this.temperature = fields?.temperature ?? this.temperature;
+    if (this.temperature && (this.temperature < 0 || this.temperature > 1)) {
+      throw new Error("`temperature` must be in the range of [0.0,1.0]");
+    }
+
+    this.topP = fields?.topP ?? this.topP;
+    if (this.topP && this.topP < 0) {
+      throw new Error("`topP` must be a positive integer");
+    }
+
+    this.topK = fields?.topK ?? this.topK;
+    if (this.topK && this.topK < 0) {
+      throw new Error("`topK` must be a positive integer");
+    }
+
+    this.examples = fields?.examples ?? this.examples;
+
+    this.apiKey =
+      fields?.apiKey ?? getEnvironmentVariable("GOOGLEPALM_API_KEY");
+    if (!this.apiKey) {
+      throw new Error(
+        "Please set an API key for Google Palm 2 in the environment variable GOOGLEPALM_API_KEY or in the `apiKey` field of the GooglePalm constructor"
+      );
+    }
+
+    this.client = new DiscussServiceClient({
+      authClient: new GoogleAuth().fromAPIKey(this.apiKey),
+    });
+  }
+
+  _combineLLMOutput() {
+    return [];
+  }
+
+  _llmType() {
+    return "googlepalm";
+  }
+
+  async _generate(
+    messages: BaseMessage[],
+    options: this["ParsedCallOptions"],
+    runManager?: CallbackManagerForLLMRun
+  ): Promise<ChatResult> {
+    const chatResult = await this.caller.callWithOptions(
+      { signal: options.signal },
+      this._palmGenerateMessage.bind(this),
+      messages
+    );
+
+    // Google Palm doesn't provide streaming as of now. But to support streaming handlers
+    // we call the handler with entire response text
+    void runManager?.handleLLMNewToken(
+      chatResult.generations.length > 0 ? chatResult.generations[0].text : ""
+    );
+
+    return chatResult;
+  }
+
+  async _palmGenerateMessage(messages: BaseMessage[]): Promise<ChatResult> {
+    const res = await this.client.generateMessage({
+      candidateCount: 1,
+      model: this.model,
+      temperature: this.temperature,
+      topK: this.topK,
+      topP: this.topP,
+      prompt: {
+        context: this._getPalmContextInstruction(messages),
+        examples: this.examples,
+        messages: this._mapBaseMessagesToPalmMessages(messages),
+      },
+    });
+    return this._mapPalmMessagesToChatResult(res[0]);
+  }
+
+  _getPalmContextInstruction(messages: BaseMessage[]): string | undefined {
+    // filters out all 'system' messages, but select the first message
```

Comment:
I can't access the docs, but Vertex AI accepts a system message only as the first message.

This is different for PaLM?

Reply:
Example documentation [here](https://developers.generativeai.google/tutorials/chat_node_quickstart#generate_messages). PaLM `DiscussService` client needs a `context` (which corresponds to `SystemMessage`) and and a array of `messages` (which corresponds to `HumanMessage` & `AIMessage`).

```javascript
const result = await client.generateMessage({
    ...
    prompt: {
      // optional, preamble context to prime responses
      context: "Respond to all questions with a rhyming poem.",
      examples: [
         ...
      ],
      // Required. Alternating prompt/response messages.
      messages: [{ content: "How tall is the Eiffel Tower?" }],
    },
  });
```

Since it's a convention to keep `SystemMessage` as always the first message in `langchain`, let me adopt the same :)